### ISSUE TITLE
Default Realm not updating quickly enough causing TableView assertion to trigger. 

### ISSUE THREAD

**Issue by ArEnSc** [realm/realm-cocoa#2912]
Hi, I am performing a delete operation on an object inside a realm, on a serial background thread.

I am performing the after a function callback is completed synchronously.

```objective-c

// delete realm object on separate thread using id to retrieve the realm object
// completed synchronously...
 [self.db saveObject:t withCompleteBlock:^(BOOL complete) {
       
        [weakSelf.tableView beginUpdates];
            [weakSelf.tableView deleteRowsAtIndexPaths:self.indexPathes withRowAnimation:UITableViewRowAnimationLeft];
            [weakSelf.tableView endUpdates];// Crashes here
        
   }];


          
```

Upon checking how many objects are left inside the RLMResults array, it concludes that the object was not removed between begin and end updates.

So why does realm not delete the object right away atomically ?

The reason is crashes is, that the tableView expects count of the tasks

**Comment by ArEnSc** [comment]
Sample code will come later.

The solution is to use the notification

```objective-c
 self.token = [RLMRealm.defaultRealm addNotificationBlock:^(NSString * _Nonnull notification, RLMRealm * _Nonnull realm) {
    
        if (weakSelf.indexPathes !=nil) {
            [weakSelf.tableView beginUpdates];
            [weakSelf.tableView deleteRowsAtIndexPaths:self.indexPathes withRowAnimation:UITableViewRowAnimationLeft];
            [weakSelf.tableView endUpdates];
            weakSelf.indexPathes = nil;
        }
        else {
            [weakSelf.tableView reloadData];
        }
    }];


```

**Comment by kishikawakatsumi** [comment]
@ArEnSc Calling `[defaultRealm refresh]` manually before tableView's `beginUpdates` method might fix this issue?


To avoid having data change beneath your feet in the middle of a function, we only update which version of the data an RLMRealm uses between each iteration of the runloop. Unfortunately, because GCD blocks are the highest-priority events, dispatch_async(dispatch_get_main_queue(), ^{ ... }) immediately after a commit can result in the block being run on the main thread before our autorefresh signal, resulting in the block seeing the version of the Realm from before the commit.

You can work around this by explicitly calling [realm refresh] at the beginning of your block that needs to see the latest data. The more proper way would be to instead do the work in a Realm notification block rather than a dispatched block, but with the current limited notifications that'd be difficult to do.

**Comment by ArEnSc** [comment]
Ahhh I understand I was thinking that as well, I called refresh in the background thread instead of before the begin updates and end updates

**Comment by ArEnSc** [comment]
[realm refresh] returns true, yet it still doesn't refresh the current view of the realm.

btw I am using a serial queue

```objective-c



```

**Comment by bdash** [comment]
Can you share the implementation of your `-saveObject:withCompleteBlock:` method? That should help us understand why calling `-refresh` in your block isn't sufficient.

**Comment by ArEnSc** [comment]
```objective-c

-(void) saveObject:(BaseModel *)baseModel withCompleteBlock:(RealmOperation) complete {
    
    __block NSString* objectId;
    if(baseModel.markForDelete == YES) {
        objectId = baseModel.Id;
    }
    dispatch_async(self.queue, ^{
       
        if (objectId !=nil) {
            Task* updateOrDeleteModel = [Task objectForPrimaryKey:objectId];
            updateOrDeleteModel.markForDelete = YES;
            
            NSAssert(updateOrDeleteModel !=nil, @"Equal to Nil?");
            [self saveObject:updateOrDeleteModel];
        }
        else {
            [self saveObject:baseModel];
        }
        
        // Realm doesn't delete objects fast enough.
        dispatch_sync(dispatch_get_main_queue(), ^{
            complete(YES);
        });
    
    });
    
}


```

**Comment by bdash** [comment]
Hmm. Nothing seems obviously wrong with that code. The call to `dispatch_sync` should probably be a `dispatch_async` since there doesn't appear to be any reason to tie up the background queue while calling the completion block, but that doesn't seem like it'd cause the problem you're seeing.

In the initial code snippet you provided, I see that it's using `self.indexPathes` when calling into the table view. It sounds like your problem is related to it having an incorrect value, but I don't see anywhere in the code you've provided where that value is set. Can you share more about what `self.indexPathes` is and how it's used?

Would you also mind sharing the backtrace of the crash you're seeing, along with the exception message if the crash is due to an exception being thrown?

**Comment by jpsim** [comment]
@ArEnSc just following up on @bdash's last comment. Could you please share more about what `self.indexPathes` is and how it's used? Also useful would be the backtrace of the crash you're seeing, along with the exception message if the crash is due to an exception being thrown.

**Comment by ArEnSc** [comment]
Hey sorry I have been busy!, right so self.indexPathes does retain the index paths let me quickly create a demo project tomorrow to illustrate the issue, refresh is not working for me.

**Comment by siuying** [comment]
I also happens to found updates made in a thread not accessible from the other. I wonder do I still need a refresh() if I create a new realm object from Realm(config:)?

**Comment by jpsim** [comment]
> I also happens to found updates made in a thread sometimes inaccessible from the other. I wonder do I still need a refresh() if I create a new realm object from Realm(config:) everytime I access them?

You should never expect changes made in one thread to be immediately available in another. It might work sometimes, but we offer no guarantees. That's what `-[RLMRealm refresh]` is for.

I strongly encourage you to read our [documentation on Threading](https://realm.io/docs/objc/latest/#threading), which aims to clearly describe what our thread availability guarantees are.

If there's more information you'd like to see in there, please let us know and we'll be sure to consider it!

**Comment by ArEnSc** [comment]
What would solve this problem, is if we could pass a note through for a transaction and then have it be recieved by the listener, my work around to send a notification to a queue every time an action occurs on the ui then process the notification in the listener to update the UI.
However this is not thread safe, because we cannot know ahead of time whether or not if a delete finished or an update finished, so it means I have to use a serial queue.

```objective-c
  self.token = [RLMRealm.defaultRealm addNotificationBlock:^(NSString * _Nonnull notification, RLMRealm * _Nonnull realm) {
        NSLog(@"Notification!");
        RealmUINotification* note = [weakSelf.center pullNotification];
        NSArray* indicies = [note indicies];
        
        [weakSelf.tableView beginUpdates];
        
        if ([note noteType] == NoteTypeCreate) {
            [weakSelf.tableView insertRowsAtIndexPaths:indicies withRowAnimation:UITableViewRowAnimationTop];
        }
        else if ([note noteType] == NoteTypeRetrieve) {
            [weakSelf.tableView reloadData];
        }
        else if ([note noteType] == NoteTypeUpdate) {
            [weakSelf.tableView reloadRowsAtIndexPaths:indicies withRowAnimation:UITableViewRowAnimationBottom];
        }
        else if ([note noteType] == NoteTypeDelete) {
            [weakSelf.tableView deleteRowsAtIndexPaths:indicies withRowAnimation:UITableViewRowAnimationLeft];
        }
        
        [weakSelf.tableView endUpdates];
        
    }];



```

**Comment by jpsim** [comment]
> What would solve this problem, is if we could pass a note through for a transaction and then have it be recieved by the listener, my work around to send a notification to a queue every time an action occurs on the ui then process the notification in the listener to update the UI.

We're aiming to solve this in a more general way with #601. Until then, posting a notification should be acceptable, but if this is being done across threads, be sure to `-[RLMRealm refresh]` to have the latest data.

> However this is not thread safe, because we cannot know ahead of time whether or not if a delete finished or an update finished, so it means I have to use a serial queue.

Why not include that information in your notification?

**Comment by ArEnSc** [comment]
My notification works from the UI End I don't want to throw in the notification system into the backend system perhaps I should do it that way. but that means if were to do it asynchrnously it would have lock the Notiication centre.

**Comment by jpsim** [comment]
> My notification works from the UI End I don't want to throw in the notification system into the backend system perhaps I should do it that way. but that means if were to do it asynchrnously it would have lock the Notiication centre.

This is one of the situations that's difficult to guard against without proper support for #601. For example, it's possible that Realm update notifications be coalesced for multiple write transactions. Or if there are multiple places where you need to respond to a change notification.

One complex and ugly solution would be to inline change notifications as Realm objects within your write transactions, which would guarantee that you get them at the same time as the updates themselves.

I'm sorry I can't offer you a better workaround, but if one existed, we wouldn't need #601 :disappointed: .

**Comment by ArEnSc** [comment]
When do you think that #601 will be ready, I am in the process of writing a book for realm right now, that clarifies a lot of the stuff left out of the documentation.

**Comment by jpsim** [comment]
> When do you think that #601 will be ready?

It's our next big task once #2842 lands, but we can't provide a timeline at the moment.

> I am in the process of writing a book for realm right now, that clarifies a lot of the stuff left out of the documentation.

On the one hand, it's great that you're writing a book about Realm! On the other, if there are parts of our documentation that aren't clear, please let us know! We'll do our best to improve it.

**Comment by siuying** [comment]
@jpsim thanks. I just want to confirm, should we do a refresh even we are creating a new instance of realm? Because one might assume a new realm object is up to date.

If refresh is required, maybe the examples in the threading section should be updated?

**Comment by jpsim** [comment]
No. New Realm instances are always up to date. This is explained in the "Threading" section of our docs.

**Comment by siuying** [comment]
@jpsim is it possible that a cached realm being returned, if I create a new Realm? Because if I added refresh() on every creation inside thread, the bugs I'm seeing (and spending days to debug) are gone.

**Comment by jpsim** [comment]
Sorry for the delay, the holidays got in the way :christmas_tree:.

> is it possible that a cached realm being returned, if I create a new Realm? Because if I added refresh() on every creation inside thread, the bugs I'm seeing (and spending days to debug) are gone.

Yes, Realms are cached per-thread, so if you're on a thread (thread "A") within one runloop iteration (or on a thread with no runloop), open a realm, then a different thread (thread "B") commits a change, then you open a new Realm on thread "A", that new Realm is the same as the first one from thread "A" which was cached, so you'd need to call `-refresh` if you wanted to see the changes from the write transaction from thread "B".

I suppose we could clarify this further in our documentation by saying "When you initially open a Realm on a thread" rather than just "When you initially open a Realm": https://realm.io/docs/objc/latest/#seeing-changes-from-other-threads

**Comment by jpsim** [comment]
@siuying @ArEnSc I've clarified this in the docs and will now close this issue because I think I've addressed all your questions. If you have any further questions related to this topic, you may re-open the issue. For all other questions, I encourage you to file a new issue on this repo and we'll be happy to help!

**Comment by ArEnSc** [comment]
Yeah I figure you will eventually need to get custom notifications working inorder to solve this problem.
